File: LeetCode/Medium/0079-word-search/0079-word-search.py
class Solution:

    def DFSHelper(self, board, i, j, word, word_idx, prev_matrix):
        r, c = len(board), len(board[0])
        
        if 0<=i<r and 0<=j<c and board[i][j] == word[word_idx] and prev_matrix[i][j] == 1:
            if word_idx == len(word)-1: 
                return 1
            
            prev_matrix[i][j] = 0   
            
            result = self.DFSHelper(board, i, j+1, word, word_idx+1, prev_matrix) + \
                     self.DFSHelper(board, i+1, j, word, word_idx+1, prev_matrix) + \
                     self.DFSHelper(board, i, j-1, word, word_idx+1, prev_matrix) + \
                     self.DFSHelper(board, i-1, j, word, word_idx+1, prev_matrix)
            
            prev_matrix[i][j] = 1

            return result 
            
        return 0

    def exist(self, board: List[List[str]], word: str) -> bool:

        r, c = len(board), len(board[0])

        for m in range(r):
            for n in range(c):
                prev_matrix = [[1 for _ in range(c+1)] for _ in range(r+1)]
                if self.DFSHelper(board, m, n, word, 0, prev_matrix):
                    return True
                
        return False
                    

                
        

File: LeetCode/Medium/0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        unordered_map<int, bool> visited;
        unordered_map<char, int> cnt;
        
				// map에 word 각 글자가 몇개 나오는지 저장
        for (auto c : word) {
            cnt[c] += 1;
        }
				        
				// maxVal = cnt에서 가장 value가 큰 key-value
        auto maxVal = max_element(cnt.begin(), cnt.end(), 
											[](const auto &x, const auto &y) {return x.second < y.second;});
			
				// word 첫글자가 가장 value큰 값이면 단어 뒤집기??
        if (cnt[word[0]] == cnt[maxVal->first]) {
            reverse(word.begin(), word.end());
        }
			
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (exist(board, word, i, j, 0, visited))
                    return true;
            }
        }
        
        return false;
    }
private:
    bool exist(vector<vector<char>>& board, string& word, int row, int col, int idx, unordered_map<int, bool> visited) {
        if (visited[row*10+col]) return false;
        if (row >= board.size() || col >= board[0].size() || row < 0 || col < 0) return false;
        else if (board[row][col] == word[idx]) {
            if (idx == word.size() - 1) 
                return true;
            else {
                visited[row*10 + col] = true;
                return exist(board, word, row - 1, col, idx+1, visited) 
                        || exist(board, word, row + 1, col, idx+1, visited)
                        || exist(board, word, row, col - 1, idx+1, visited) 
                        || exist(board, word, row, col + 1, idx+1, visited);
            }     
        }
        else return false;
    }
};